imap_processing/cdf/utils.py

```diff
@@ -16,39 +16,49 @@
 logger = logging.getLogger(__name__)
 
 
-def calc_start_time(
-    shcoarse_time: float,
-    launch_time: Optional[np.datetime64] = imap_processing.launch_time,
-) -> np.datetime64:
-    """
-    Calculate the datetime64 from the CCSDS secondary header information.
+# Reference start time (launch time or epoch)
+# DEFAULT_EPOCH = np.datetime64("2010-01-01T00:01:06.184", "ns")
+IMAP_EPOCH = np.datetime64("2010-01-01T00:00:00", "ns")
+J2000_EPOCH = np.datetime64("2000-01-01T11:58:55.816", "ns")
+
 
-    Since all instrument has SHCOARSE or MET seconds, we need convert it to
-    UTC. Took this from IDEX code.
+def met_to_j2000ns(
+    met: np.typing.ArrayLike,
+    reference_epoch: Optional[np.datetime64] = IMAP_EPOCH,
+) -> np.typing.ArrayLike:
+    """
+    Convert mission elapsed time (MET) to nanoseconds from J2000.
 
     Parameters
     ----------
-    shcoarse_time : float
-        Number of seconds since epoch (nominally the launch time).
-    launch_time : np.datetime64
-        The time of launch to use as the baseline.
+    met : array_like
+        Number of seconds since epoch according to the spacecraft clock.
+    reference_epoch : np.datetime64
+        The time of reference for the mission elapsed time. The standard
+        reference time for IMAP is January 1, 2010 00:00:00 UTC. Per APL's
+        IMAP Timekeeping System Design document.
 
     Returns
     -------
-    np.timedelta64
-        The time of the event.
+    array_like or scalar, int64
+        The mission elapsed time converted to nanoseconds since the J2000 epoch.
 
     Notes
     -----
-    TODO - move this into imap-data-access? How should it be used?
-    This conversion is temporary for now, and will need SPICE in the future.
-    Nick Dutton mentioned that s/c clock start epoch is
-        jan-1-2010-00:01:06.184 ET
-    We will use this for now.
+    This conversion is temporary for now, and will need SPICE in the future to
+    account for spacecraft clock drift.
     """
-    # Get the datetime of Jan 1 2010 as the start date
-    time_delta = np.timedelta64(int(shcoarse_time * 1e9), "ns")
-    return launch_time + time_delta
+    # Mission elapsed time is in seconds, convert to nanoseconds
+    # NOTE: We need to multiply the incoming met by 1e9 first because we could have
+    #       float input and we want to keep ns precision in those floats
+    # NOTE: We need int64 here when running on 32bit systems as plain int will default
+    #       to 32bit and overflow due to the nanosecond multiplication
+    time_array = (np.asarray(met) * 1e9).astype(np.int64)
```

Comment:
It appears that the multiplication will overflow if `met` is passed in as a 32 bit integer (signed or unsigned). I see that SHCOARSE is a `uint32` in packets. Doing the math on seconds since the IMAP_EPOCH:
```
# 15 years * 365.25 days/year * 24 hours/day * 60 minutes/hour * 60 seconds/minute * 1e9 ns/second
np.log2(15 * 365.25 * 24 * 60 * 60 * 1e9) = 58.7
```

Reply:
Very good catch! I was getting lucky before as numpy did the cast to float automatically for us.
```
np.asarray(np.int32(2**30)) -> np.int32
np.asarray(np.int32(2**30)) * 1e9 -> np.float
```
I believe that is changing in numpy2.0 so I've updated it to be explicit by adding in `np.asarray(met, dtype=float)`.